### DISCUSSION TITLE
substructure counter: Knime outcome different than Python

### DISCUSSION THREAD

**Discussion by DocMinus** [rdkit/rdkit#5921] [Q&A]
(rephrasing this question since I might have posed it originally as two intertwined questions?)
**What** - attempting to count total number all atom-atom in a molecule using GetSubstructMatches, comparing result versus a Workflow in Knime using the Rdkit nodes. (I know there is a Knime forum, but that isn't as regularly visited by mainter?)

**Problem** - The counts using the same smarts give a different outcome. Meaning there must be something different happening under the hood? 
Here is a **code snippet** (only selected fragments) that I tested in Python 3.8+, Rdkit V2022.09.1. I am aware (especially from previous answer) that one could/should write the smarts different, but it doesn't explain the outcome. 
```
from rdkit import Chem

smiles = "CC1=CC=C(CNC(N)C2=CC=CC(CO)=C2CC=C)N=C1"
#smiles = "C1=CC=CC=C1"
mol = Chem.MolFromSmiles(smiles)
mol = Chem.AddHs(mol)  # doesn't seem to matter?

frags = ["CC","CN","C=C", "[C;H2]"]
_match = []
for group in frags: 
    _match.append(len(mol.GetSubstructMatches(Chem.MolFromSmarts(group), uniquify=True)))
matches = dict(zip(frags, _match))
print(matches)
```
_The output for  benzene gives:_
{'CC': 0, 'CN': 0, 'C=C': 0, '[C;H2]': 0} 

_The other compound gives:_
{'CC': 1, 'CN': 3, 'C=C': 2, '[C;H2]': 4}

Now look below, **Rdkit in Knime** gives a completely different result with the same frags (workflow not included here; I can mention though that I used it with/without kekulizer/aromatizer node before the matcher):
_Benzene:_
CC:   3, CN: 0, C=C: 3, C;H2: 0 
_Other compound_:
CC: 11, CN: 6, C=C: 1, C;H2:4   

Anyone thoughts?

**Comment by bp-kelley** [comment]
C in smarts is an aliphatic carbon, c is aromatic.  To count any carbon you need to use [#6]

It might be worthwhile to read daylight’s smarts guide:  

https://www.daylight.com/dayhtml/doc/theory/theory.smarts.html

**Reply by DocMinus** [reply]
Yes, you are of course correct, I should have remembered, considering having done it in other contexts.  doh me!
**But**: doesn't explain why Knime node gives a different answer than the Python code?

**Reply by DocMinus** [reply]
PS: rewrote my question to be more Knime/Python comparison centric. Hope that's ok.

**Comment by bp-kelley** [comment]
The knime result is a puzzle to be sure,  but it acts like it is substructure is kind of matching on the kekule form of the molecule.

```
smiles = "CC1=CC=C(CNC(N)C2=CC=CC(CO)=C2CC=C)N=C1"
mol = Chem.MolFromSmiles(smiles, sanitize=False)
Chem.Kekulize(mol)

frags = ["CC","CN","C=C", "[C;H2]"]
_match = []
for group in frags: 
    _match.append(len(mol.GetSubstructMatches(Chem.MolFromSmarts(group), uniquify=True)))
matches = dict(zip(frags, _match))
print(matches)
```

```
{'CC': 11, 'CN': 4, 'C=C': 6, '[C;H2]': 4}
```

**Reply by bp-kelley** [reply]
But this doesn't match the C=C fragment.

**Comment by kienerj** [comment]
Can you post your KNIME workflow? I tried to repeat and actually get yet another result:

![image](https://user-images.githubusercontent.com/1567309/211270379-6f724b85-5945-4320-9e0d-b8a26bde4859.png)

C=C is 1 which I think is actually the correct result over what you get in python code (2)?

**Reply by DocMinus** [reply]
thanks for looking more into this. here is a screenshot. Actually, 3 different ways of doing it, 2 different results.
top left: aromatized form with  hydrogens, same as kekulized form without Hs. But, Aromatized without H gives 0 matches...
![knime](https://user-images.githubusercontent.com/44326879/212273318-04364618-8ca0-4884-8b2d-eb53b2ab6f4a.png)
